libavcodec/vvc/dec.c

```diff
@@ -887,16 +999,192 @@ static int set_output_format(const VVCContext *s, const AVFrame *output)
+        w = s->avctx->coded_width  >> sps->hshift[i];
+        h = s->avctx->coded_height >> sps->vshift[i];
+
+        checksum = 0;
+        for (y = 0; y < h; y++) {
+            const uint8_t *src = frame->data[i] + y * frame->linesize[i];
+            for (x = 0; x < w; x++) {
+                int xor_mask = (x & 0xFF) ^ (y & 0xFF) ^ (x >> 8) ^ (y >> 8);
+                checksum = CAL_CHECKSUM(src[x << sps->pixel_shift]);
+                if (sps->pixel_shift)
+                    checksum = CAL_CHECKSUM(src[(x << sps->pixel_shift) + 1]);
+            }
+        }
+
+        av_log(s->avctx, checksum == fc->sei.picture_hash.checksum[i] ? AV_LOG_DEBUG : AV_LOG_ERROR ,
+            "Verifying CRC for frame with POC %d ", fc->ref->poc);
+        if (checksum == fc->sei.picture_hash.checksum[i]) {
+            av_log(s->avctx, AV_LOG_DEBUG, "plane %d - correct - %d\n",
+                i, checksum);
+        } else {
+            av_log(s->avctx, AV_LOG_ERROR, "mismatching CRC of plane %d - %d != %d\n",
+                i, checksum, fc->sei.picture_hash.checksum[i]);
+            return AVERROR_INVALIDDATA;
+        }
+    }
+
+    return 0;
+}
+
+static int frame_end(VVCContext *s, VVCFrameContext *fc)
+{
+    const AVFilmGrainParams *fgp;
+    int ret = 0;
+
+    if (fc->ref->needs_fg) {
+        av_assert0(fc->ref->frame_grain->buf[0]);
+        fgp = av_film_grain_params_select(fc->ref->frame);
+        switch (fgp->type) {
+        case AV_FILM_GRAIN_PARAMS_NONE:
+            av_assert0(0);
+            return AVERROR_BUG;
+        case AV_FILM_GRAIN_PARAMS_H274:
+            ret = ff_h274_apply_film_grain(fc->ref->frame_grain, fc->ref->frame,
+                &s->h274db, fgp);
+            break;
+        case AV_FILM_GRAIN_PARAMS_AV1:
+            ret = ff_aom_apply_film_grain(fc->ref->frame_grain, fc->ref->frame, fgp);
+            break;
+        }
+    }
+
+    if (!s->avctx->hwaccel && s->avctx->err_recognition & AV_EF_CRCCHECK) {
+        if (fc->sei.picture_hash.present) {
+            if (fc->sei.picture_hash.hash_type == 0)
+                ret = verify_md5(s, fc);
+            else if (fc->sei.picture_hash.hash_type == 1)
+                ret = verify_crc(s, fc);
+            else if (fc->sei.picture_hash.hash_type == 2)
+                ret = verify_checksum(s, fc);
+            if (ret < 0 && s->avctx->err_recognition & AV_EF_EXPLODE)
+                return ret;
+        }
+    }
+
+    return 0;
+}
+
 static int wait_delayed_frame(VVCContext *s, AVFrame *output, int *got_output)
 {
     VVCFrameContext *delayed = get_frame_context(s, s->fcs, s->nb_frames - s->nb_delayed);
     int ret                  = ff_vvc_frame_wait(s, delayed);
 
-    if (!ret && delayed->output_frame->buf[0] && output) {
-        av_frame_move_ref(output, delayed->output_frame);
-        ret = set_output_format(s, output);
-        if (!ret)
-            *got_output = 1;
+    if (!ret) {
+        ret = frame_end(s, delayed);
```

Comment:
do we need to call it export_sei and move it to "  if (delayed->output_frame->buf[0] && output) {"

Reply:
Maybe not. We already have a function `set_side_data` for export sei.

Reply:
I tried add it to `if (delayed->output_frame->buf[0] && output) {` at my first version but it doesn't work.

Reply:
Could you have a check?